### PR TITLE
Implement API usage metrics for gce storage

### PR DESCRIPTION
**Pull request by gnufied**
**What this PR does / why we need it**:

This PR implements support for emitting metrics from GCE about storage operations.

**Which issue this PR fixes** 

Fixes https://github.com/kubernetes/features/issues/182

**Release note**:
```
Add support for emitting metrics from GCE cloudprovider about storage operations.
```

### CODE REVIEW

File: pkg/cloudprovider/providers/gce/gce.go
Comment (sttts) [inline pkg/cloudprovider/providers/gce/gce.go]: why public?

File: pkg/cloudprovider/providers/gce/gce.go
Comment (sttts) [inline pkg/cloudprovider/providers/gce/gce.go]: IMO ugly indention

File: pkg/cloudprovider/providers/gce/gce.go
Comment (sttts) [inline pkg/cloudprovider/providers/gce/gce.go]: why public?

File: pkg/cloudprovider/providers/gce/gce.go
Comment (sttts) [inline pkg/cloudprovider/providers/gce/gce.go]:
```
if !ok {
  return nil
}
```

Comment (sttts) [inline pkg/cloudprovider/providers/gce/gce.go]: return func(....)

File: pkg/cloudprovider/providers/gce/gce.go
Comment (sttts) [inline pkg/cloudprovider/providers/gce/gce.go]: are we sure that `gceMetricMap[apiNamespace.apiCall]` is never nil?

File: pkg/cloudprovider/providers/gce/gce.go
Comment (sttts) [inline pkg/cloudprovider/providers/gce/gce.go]: microsecondsSince

File: pkg/cloudprovider/providers/gce/gce.go
Comment (sttts) [inline pkg/cloudprovider/providers/gce/gce.go]: no need for empty line above? And why not merge the non-kube blocks?

Comment (gnufied) [inline pkg/cloudprovider/providers/gce/gce.go]: Thank you for reviewing this. I pushed this as a WIP and used it as a starting point for [CloudProvider metrics](https://github.com/kubernetes/features/issues/182) proposal. I will clean it up and add more checks for nil and stuff. 

I will give you another shout when it is ready for review.

Comment (sttts) [inline pkg/cloudprovider/providers/gce/gce.go]: Ok.

**Comment by gnufied** [comment]
@sttts I have fixed the review comments and this PR is ready for review. Also adding @piosz @jingxu97

**Comment by jingxu97** [comment]
cc @bowei

**Comment by piosz** [comment]
LGTM
cc @fabxc @brancz

File: pkg/cloudprovider/providers/gce/gce_metrics.go
Comment (piosz) [inline pkg/cloudprovider/providers/gce/gce_metrics.go]: 2017

File: pkg/cloudprovider/providers/gce/gce_metrics.go
Comment (brancz) [inline pkg/cloudprovider/providers/gce/gce_metrics.go]: note the naming conventions for histograms: https://prometheus.io/docs/practices/histograms/

`gce_instance_list_duration_seconds` would be applicable here. Note that metrics are also meant to be in the base unit. Generally following the According to the [instrumentation guide](https://github.com/kubernetes/community/blob/master/contributors/devel/instrumentation.md).

Same applies to all other metrics introduced here.

File: pkg/cloudprovider/providers/gce/gce_metrics.go
Comment (brancz) [inline pkg/cloudprovider/providers/gce/gce_metrics.go]: According to the [instrumentation guide](https://github.com/kubernetes/community/blob/master/contributors/devel/instrumentation.md), this should be just `namespace`.

**Review by brancz — CHANGES_REQUESTED** [review]

File: pkg/cloudprovider/providers/gce/gce_metrics.go
Comment (gnufied) [inline pkg/cloudprovider/providers/gce/gce_metrics.go]: Do you mean - these metrics should only be collected in seconds? not in ms as done here? 

I couldn't find any info. about that in linked documentation

Comment (brancz) [inline pkg/cloudprovider/providers/gce/gce_metrics.go]: Sorry, pointed at the wrong resources there, this is what I meant: https://prometheus.io/docs/practices/naming/#metric-names

File: pkg/cloudprovider/providers/gce/gce_metrics.go
Comment (gnufied) [inline pkg/cloudprovider/providers/gce/gce_metrics.go]: fixed

File: pkg/cloudprovider/providers/gce/gce_metrics.go
Comment (gnufied) [inline pkg/cloudprovider/providers/gce/gce_metrics.go]: fixed.

**Comment by gnufied** [comment]
@brancz fixed all review comments. ptal

**Comment by brancz** [comment]
lgtm :+1:

**Comment by gnufied** [comment]
@brancz @piosz someone needs to "/lgtm" and "/approve" for this PR to get merged.

cc @saad-ali

File: pkg/cloudprovider/providers/gce/gce.go
Comment (rootfs) [inline pkg/cloudprovider/providers/gce/gce.go]: are seconds too high for latency?

Comment (gnufied) [inline pkg/cloudprovider/providers/gce/gce.go]: According to https://prometheus.io/docs/practices/naming/#metric-names metrics should use base units. I guess we should be okay, since we are reporting seconds as floats anyways.

**Pull request merged**
